### PR TITLE
User Story 3 Topics

### PR DESCRIPTION
**Pull request by TKChattoraj**
### Commit fa655f7

creates the Topics CRUD
### Commit cd71e11

configures mailer through SENDGRID
### Commit ab3c2f5

revises the Topic index and show views
### Note:

This pull request does not show bookmarks since the bookmarks themselves have not yet been created.  For now the applicable views have placeholders for the bookmarks.
### Note:

Also, when the bookmark views are created they will be put into a partial, which will be used under the Topics index and show views.  Similarly the erb for the Topics index will potentially be made into its own partial to get all of that out of the views themselves.

### CODE REVIEW

File: app/models/bookmark.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+class Bookmark < ActiveRecord::Base
+  belongs_to :topic
+
+  validates :url, presence: true
```
Comment (bjblock) [inline app/models/bookmark.rb:4]: can you think of any other relevant/important validations?

File: app/views/topics/create.html.erb [hunk 2]
```diff
@@ -0,0 +1,2 @@
+<h1>Topics#create</h1>
+<p>Find me in app/views/topics/create.html.erb</p>
```
Comment (bjblock) [inline app/views/topics/create.html.erb:2]: is this file needed?
which other view files are or are not needed?

File: app/views/topics/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+<h1>New Topic</h1>
+
+<div class="row">
+  <div class="col-md-4">
+
+  </div>
+  <div class="col-md-8">
+    <%= form_for @topic do |f| %>
```
Comment (bjblock) [inline app/views/topics/new.html.erb:8]: the `topic_params` in the `topics_controller` appear to have a potential parameter `:user_id`, but I don't see it in the form.  when/where will that attribute get assigned?

File: app/views/topics/show.html.erb [hunk 4]
```diff
@@ -0,0 +1,19 @@
+<div class="container-fluid">
+<h1 class="text-center">Topic:  <%= @topic.title%></h1>
+
+
+
+<div class="row">
+  <% 7.times do |n|%>
```
Comment (bjblock) [inline app/views/topics/show.html.erb:7]: code like this is called "Technical Debt", meaning it's code that we know is problematic and will have to be removed later.  although it's preferable to never introduce technical debt into the codebase, it's also usually unavoidable to be 100% technical debt free.  so, when necessary, at least identify (with a comment) technical debt that will be addressed later.

File: db/seeds.rb [hunk 5]
```diff
@@ -1,7 +1,22 @@
-# This file should contain all the record creation needed to seed the database with its default values.
-# The data can then be loaded with the rake db:seed (or created alongside the db with db:setup).
-#
-# Examples:
-#
-#   cities = City.create([{ name: 'Chicago' }, { name: 'Copenhagen' }])
-#   Mayor.create(name: 'Emanuel', city: cities.first)
+require 'factory_girl_rails'
```
Comment (bjblock) [inline db/seeds.rb:1]: is this being used?

File: spec/factories/bookmarks.rb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+include RandomData
```
Comment (bjblock) [inline spec/factories/bookmarks.rb:1]: is this being used?  (and also in spec/factories/topics.rb)

File: app/views/topics/new.html.erb (hunk 3 above)
Comment (TKChattoraj) [inline app/views/topics/new.html.erb:8]: restructured topics to be nested under user for new/create, edit/update and delete/destroy.  When creating a topic, the user is automatically supplied now.

File: app/models/bookmark.rb (hunk 1 above)
Comment (TKChattoraj) [inline app/models/bookmark.rb:4]: yes-revised to validate `url` for uniqueness and format.  Use the following to validate formate:  
`URL_REGEX = /\A((https?:\/\/)|(www.))[\w]+\.com\z/i`

I seemed to remember that we briefly discussed another way to validate format instead of through REGEX. Am I imagining that?

File: app/views/topics/create.html.erb (hunk 2 above)
Comment (TKChattoraj) [inline app/views/topics/create.html.erb:2]: I am not sure as I don't really know how Topics is going to flesh out.  I kept the Topic  create, destroy, edit and update views as I am not sure whether things will change that will require needing them.  For example--in the end are only Admins going to be allowed to create topics, edit or destroy them?  If so, then perhaps separate views for these are needed.  Maybe the philosophy is to not retain stuff in case you need it, but create it at the time you do.  Let's talk about that.  I think some of this can be answered by which direction the app is going?  Under what circumstances are topics created?  Who can delete a topic? Any user?  Or the user who created it.  I understand that Topics will include Bookmarks from multiple users and so can a user who created a Topic delete it even if other uses have Bookmarks under it?

File: app/views/topics/show.html.erb (hunk 4 above)
Comment (TKChattoraj) [inline app/views/topics/show.html.erb:7]: Got it.  This particular code has since been replaced.

File: db/seeds.rb (hunk 5 above)
Comment (TKChattoraj) [inline db/seeds.rb:1]: no.  I initially anticipated using it, but ended up not.

File: spec/factories/bookmarks.rb (hunk 6 above)
Comment (TKChattoraj) [inline spec/factories/bookmarks.rb:1]: no.  I originally included it, thinking that I might want to use it.

File: app/models/bookmark.rb (hunk 1 above)
Comment (bjblock) [inline app/models/bookmark.rb:4]: do bookmarks have any context if not associated to a topic?  will the app work correctly if they're ever not associated to a topic?  if either is no, then there's probably another validation that would be important.

**Comment by TKChattoraj** [comment]
all comments/questions addressed.
